gtp_engine.py

```diff
@@ -0,0 +1,197 @@
+    _print_msg("=", msg_id, msg)
+
+
+def _convert_args(handler, args):
+    """Convert a list of command arguments to types specified by the handler.
+
+    Args:
+      handler: a command handler function.
+      args: the list of string arguments to pass to handler.
+
+    Returns:
+      A new list containing `args` that have been converted to the expected type
+      for `handler`. For each function parameter of `handler` that has either an
+      explicit type annotation or a non-None default value, the corresponding
+      element in `args` is converted to that type.
+    """
+
+    args = list(args)
+    params = inspect.signature(handler).parameters
+    for i, (arg, name) in enumerate(zip(args, params)):
+        default = params[name].default
+        annotation = params[name].annotation
+
+        if annotation != inspect.Parameter.empty:
+            if isinstance(annotation, type) and annotation != str:
+                # The parameter is annotated with a type that isn't str: convert
+                # the arg to that type.
+                args[i] = annotation(arg)
+        elif default != inspect.Parameter.empty:
+            if default is not None and not isinstance(default, str):
+                # The parameter has a default value that isn't None or a str:
+                # convert the arg to the default value's type.
+                args[i] = type(default)(arg)
+
+    return args
+
+
+class Engine(object):
+    """A simple GTP engine.
+
+    The engine by itself doesn't do anything: clients must register command
+    handler objects using `add_cmd_handler`.
+    """
+
+    def __init__(self):
+        self.cmds = {}
+
+    def add_cmd_handler(self, handler_obj):
+        """Registers a new command handler object.
+
+        All methods on `handler_obj` whose name starts with "cmd_" are
+        registered as a GTP command. For example, the method cmd_genmove will
+        be invoked when the engine receives a genmove command.
+
+        Args:
+          handler_obj: the handler object to register.
+        """
+        for field in dir(handler_obj):
+            if field.startswith("cmd_"):
+                cmd = field[4:]
+                fn = getattr(handler_obj, field)
+                self.cmds[cmd] = fn
+
+    def handle_msg(self, msg):
+        msg_id, cmd, args = _parse(_preprocess(msg))
+        if not cmd:
+            # Ignore empty lines.
+            return True
+
+        if cmd == "quit":
+            _print_success(msg_id, "")
+            return False
+
+        sanitized_cmd = cmd.replace("-", "_")
+        if sanitized_cmd not in self.cmds:
+            _print_error(msg_id, "unknown command")
+            return True
+
+        try:
+            handler = self.cmds[sanitized_cmd]
+            args = _convert_args(handler, args)
+            _print_success(msg_id, handler(*args))
+        except Exception as e:
+            traceback.print_exc(file=sys.stderr)
+            if cmd in _CANONICAL_ERRORS:
```

Comment:
alternatively, if you're going to catch the error and re-emit the canonical error code, you could make the errors elsewhere be more informative